### PR TITLE
Dev Container setup for CARMA Streets

### PR DESCRIPTION
**Pull request by paulbourelly999**
# PR Details
## Description
Dev Container setup for CARMA Streets

## Related Jira Key
[CDAR-389
](https://usdot-carma.atlassian.net/browse/CDAR-389)

## Motivation and Context
Setup easy to use development and integration testing environment

## How Has This Been Tested?
Locally tested both streets_service_base and streets_service_base_lanelet_aware

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have read the [**CONTRIBUTING**](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: .devcontainer/devcontainer.json
Comment (adamlm) [inline .devcontainer/devcontainer.json:1]: Can we remove commented-out configuration options? Users can look up the Dev Container schema if they are interested.

File: .devcontainer/devcontainer.json [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	// Configure tool-specific properties.
+	// "customizations": {},
+	// Configure tool-specific properties.
+	"customizations": {
+		"vscode": {
+			"extensions": [
+				"ms-vscode.cpptools",
+				"ms-vscode.cpptools-extension-pack",
+				"ms-vscode.cmake-tools",
+				"twxs.cmake",
```
Comment (adamlm) [inline .devcontainer/devcontainer.json:49]: The C/C++ Extension Pack (`ms-vscode.cpptools-extension-pack`) will install all of these together.

**Review by adamlm — CHANGES_REQUESTED** [review]

File: README.md [hunk 2]
```diff
@@ -22,7 +22,10 @@ CARMA Streets is a component of CARMA ecosystem, which enables such a coordinati
...
 
 ## Deployment
-Docker is the primary deployment mechanism to containerize one or more services. The CARMA Streets application and other major frameworks such as Kafka will run in their own separate containers. This document will be updated with a detailed Docker deployment strategy during later design phases. 
+Docker is the primary deployment mechanism to containerize one or more services. The CARMA Streets application and other major frameworks such as Kafka will run in their own separate containers. This document will be updated with a detailed Docker deployment strategy during later design phases.
```
Comment (paulbourelly999) [inline README.md:25]: Add documentation for developing new CS service.

File: .devcontainer/devcontainer.json [hunk 3]
```diff
@@ -0,0 +1,40 @@
+{
+	"name": "Existing Docker Compose (Extend)",
+
+	"dockerComposeFile": [
+		"../docker-compose.yml",
```
Comment (dan-du-car) [inline .devcontainer/devcontainer.json:5]: Why do we have two docker-compose.yml file here. Can we rename this docker-compose.yml to docker-compose-devcontainer.yml to indicate its purpose?

File: .devcontainer/devcontainer.json [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+	],
+
+	// Uncomment for Streets Service Base dev environment
+	"service": "streets_service_base",
+	// Uncomment for Streets Service Base Lanelet aware dev environment
+	// "service": "streets_service_base_lanelet_aware",
+
+	// The optional 'workspaceFolder' property is the path VS Code should open by default when
+	// connected. This is typically a file mount in .devcontainer/docker-compose.yml
+	"workspaceFolder": "/home/${localWorkspaceFolderBasename}",
```
Comment (dan-du-car) [inline .devcontainer/devcontainer.json:16]: What is the value of this localWorkspaceFolderBasename?

File: .devcontainer/devcontainer.json [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+	// Uncomment for Sensor Data Sharing Service  .
+	// "postCreateCommand": "/home/carma-streets/sensor_data_sharing_service/build.sh",
```
Comment (dan-du-car) [inline .devcontainer/devcontainer.json:19]: Is this build.sh needed?

File: .devcontainer/docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,32 @@
+version: '3.7'
+services:
+  # Uncomment for streets service base dev environment
+  streets_service_base:
+    build:
+      context: .
```
Comment (dan-du-car) [inline .devcontainer/docker-compose.yml:6]: Is this context the parent folder using `..`?

File: .devcontainer/docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+      dockerfile: ./streets_service_base/Dockerfile
+      args:
+        UBUNTU_CODENAME: bionic
+    network_mode: host
+
+  # Uncomment for streets service base lanelet aware dev environment
```
Comment (dan-du-car) [inline .devcontainer/docker-compose.yml:12]: We can only run one service at a time?

File: .devcontainer/devcontainer.json (hunk 3 above)
Comment (paulbourelly999) [reply .devcontainer/devcontainer.json:5]: Sure

File: .devcontainer/devcontainer.json (hunk 4 above)
Comment (paulbourelly999) [reply .devcontainer/devcontainer.json:16]: It evaluates to the volume setup in the docker-compose-devcontainer.yml to `/home/carma-streets`. Essentially the entire CARMA-Streets repo is copied to home/carma-streets

File: .devcontainer/docker-compose.yml (hunk 6 above)
Comment (paulbourelly999) [reply .devcontainer/docker-compose.yml:6]: The build context here is relative to the project level docker-compose file

File: .devcontainer/docker-compose.yml (hunk 7 above)
Comment (paulbourelly999) [reply .devcontainer/docker-compose.yml:12]: This is a service additional to all other carma-streets service described in the project docker-compose. Please see the added documentation

File: .devcontainer/docker-compose-devcontainer.yml [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+      - ..:/home:cached
+
+    # Uncomment the next four lines if you will use a ptrace-based debugger like C++, Go, and Rust.
+    # cap_add:
+    #   - SYS_PTRACE
+    # security_opt:
+    #   - seccomp:unconfined
+
+    # Overrides default command so things don't shut down after the process ends.
+    command: /bin/sh -c "while sleep 1000; do :; done"
```
Comment (adamlm) [inline .devcontainer/docker-compose-devcontainer.yml:31]: What is this supposed to do, and why is it necessary?

Comment (paulbourelly999) [reply .devcontainer/docker-compose-devcontainer.yml:31]: This is the command that was provided with the guided creation of a dev container environment. The comment seems to indicate that it prevents the container from shutting down on creation. We use it for our other Dev container setups (https://github.com/usdot-fhwa-OPS/V2X-Hub/blob/774b31fced876b48226302ffd093f66484a76a2f/.devcontainer/docker-compose-vscode.yml#L15).

File: .devcontainer/devcontainer.json (hunk 3 above)
Comment (dan-du-car) [inline .devcontainer/devcontainer.json:5]: Can we add comments why we need two docker-compose file?

File: .devcontainer/docker-compose-devcontainer.yml (hunk 8 above)
Comment (adamlm) [reply .devcontainer/docker-compose-devcontainer.yml:31]: I don't think this line is actually needed. We use Ubuntu's container as the base image for this. I don't see an `ENTRYPOINT` or `CMD` command in any of the derivative `Dockerfile`s, so this image should use `/bin/sh -c` as its default entry point. What are the "things" that shouldn't shutdown?

**Review by dan-du-car — APPROVED** [review]

File: .devcontainer/docker-compose-devcontainer.yml (hunk 8 above)
Comment (paulbourelly999) [reply .devcontainer/docker-compose-devcontainer.yml:31]: I can try and see if removing this works with our base images but I am against any additional exploring of what commands here work if removing this does not work. I think this is definitely something we could investigate later after more people are using the devcontainer setup. I will try removing it and confirm it work or not  and respond accordingly

Comment (paulbourelly999) [reply .devcontainer/docker-compose-devcontainer.yml:31]: It does not work without. We can investigate this later.

**Review by adamlm — APPROVED** [review]

**Pull request merged**